arch/arc/Kconfig

```diff
@@ -12,17 +12,25 @@ menu "ARC Options"
 config ARCH
 	default "arc"
 
-menu "ARC processor options"
+choice
+	prompt "ARC core family"
+	default CPU_ARCEM
 
 config	CPU_ARCEM
-	bool
-	default y
+	bool "ARC EM cores"
 	select CPU_ARCV2
 	select ATOMIC_OPERATIONS_C
 	help
 	  This option signifies the use of an ARC EM CPU
 
-endmenu
+config	CPU_ARCHS
+	bool "ARC HS cores"
+	select CPU_ARCV2
+	select ATOMIC_OPERATIONS_BUILTIN
```

Comment:
My idea was if such not very straight-forward code was never really used up until now (see we mentioned even wrong file name, https://github.com/zephyrproject-rtos/zephyr/pull/16662/commits/2f7990c8b5a012cc0484ea8ce3ab40dbc721b148) it might be a better idea to use atomic ops provided by the GCC which are known to work (even though I don't know any directed stress test for GCC's atomics) as they are used pretty often in complex software packages like Qt (and many others). At least for starters. And then once we make sure those hand-written atomics do work OK we may switch ARC HS cores to use custom version by default. That way we'll be moving with smaller steps rather than adding many major changes in 1 merge.

What if I create a new issue for switch ARC HS to custom atomics? But I guess it makes sense to do it after HS addition is merged in the master branch.

Reply:
Honestly, it's better to use the compiler builtins if they are available.

Reply:
@andrewboie well it's up to the user's philosophy. I mean in the Linux kernel and U-Boot bootloader we prefer to have as few external dependencies as possible thus trying to have everything we need from libgcc, string functions etc right in the kernel's source tree.

Moreover semantics of atomic ops might be a little bit different from project to project and so we may want to have full control on say where to put barriers etc.

But if you think in Zephyr it's better to use compiler built-ins I'm perfectly fine with that as it's much easier and allows us to reduce amount of hand-crafted assembly (which might very well have subtle issues since we're just human beings).

Reply:
> Moreover semantics of atomic ops might be a little bit different from project to project and so we may want to have full control on say where to put barriers etc.

Not sure what you mean by this.
The semantics of Zephyr's atomic APIs should be the same across arches. Otherwise this would break arch-agnostic code that uses them (futexes, etc)

> But if you think in Zephyr it's better to use compiler built-ins I'm perfectly fine with that 

For all our other arches, we prefer builtins if they exist. Arches that use the C implementations or custom assembly, AFAIK, all do this because the builtins are not implemented or not working for some reason.

Reply:
> > Moreover semantics of atomic ops might be a little bit different from project to project and so we may want to have full control on say where to put barriers etc.
> 
> Not sure what you mean by this.
> The semantics of Zephyr's atomic APIs should be the same across arches. Otherwise this would break arch-agnostic code that uses them (futexes, etc)

Indeed. I meant compiler built-in atomics might have semantics which differ from what one may expect in his particular project. Moreover I guess different compilers (not sure if that's applicable to the Zephyr today as probably only GCC is supposed to be used, but what if at some point other compilers are adopted like Clang/LLVM, proprietary ones like IAR or Keil etc) and even compiler versions might have a bit different semantics of atomics.

But probably I'm exaggerating here.
 
> > But if you think in Zephyr it's better to use compiler built-ins I'm perfectly fine with that
> 
> For all our other arches, we prefer builtins if they exist. Arches that use the C implementations or custom assembly, AFAIK, all do this because the builtins are not implemented or not working for some reason.

That's a good move and as you may see I'm all for using built-ins as more robust and simple solution. Then do you suggest to get rid of custom atomic ops for ARC at all?

Reply:
>  Then do you suggest to get rid of custom atomic ops for ARC at all?

If we can, yes.
My memory is hazy, but I think the reason why those ASM implementations exist is because GCC didn't have builtins for ARC when the ARC port was added to Zephyr, but I'd have to investigate more to be sure.

Reply:
> Indeed. I meant compiler built-in atomics might have semantics which differ from what one may expect in his particular project. Moreover I guess different compilers (not sure if that's applicable to the Zephyr today as probably only GCC is supposed to be used, but what if at some point other compilers are adopted like Clang/LLVM, proprietary ones like IAR or Keil etc) and even compiler versions might have a bit different semantics of atomics.

Ah I see. Yeah if a compiler has different atomic semantics then there would need to be a separate set of wrappers to the Zephyr atomic APIs. I expect LLVM would work the same as GCC, but who knows with other compilers.

Reply:
> 
> 
> > Then do you suggest to get rid of custom atomic ops for ARC at all?
> 
> If we can, yes.


Ok will do it shortly as otherwise we do need to have a good and thorough study of custom implementation which I'd like to escape :)

> My memory is hazy, but I think the reason why those ASM implementations exist is because GCC didn't have builtins for ARC when the ARC port was added to Zephyr, but I'd have to investigate more to be sure.

I took a look already and at least from this git repo I see ARC custom atomics were added in the very first commit 8ddf82cf70dc6f951ab477f325dee0efde3ec589. 

What's interesting since mid 2013 we were using toolchains based on GCC 4.8+ which are supposed to have atomic built-ins as we know them according to https://stackoverflow.com/a/13949440. So I'm not sure why built-ins were not used... probably they were broken? That I need to check with other people as I don't remember such kind of details as well.

Reply:
> What's interesting since mid 2013 we were using toolchains based on GCC 4.8+ which are supposed to have atomic built-ins as we know them according to stackoverflow.com/a/13949440. So I'm not sure why built-ins were not used... probably they were broken? That I need to check with other people as I don't remember such kind of details as well.

You get linker errors if the atomics aren't actually implemented for that arch.
I ran into this when I wrote the Nios II port.

Reply:
> 
> 
> > What's interesting since mid 2013 we were using toolchains based on GCC 4.8+ which are supposed to have atomic built-ins as we know them according to stackoverflow.com/a/13949440. So I'm not sure why built-ins were not used... probably they were broken? That I need to check with other people as I don't remember such kind of details as well.
> 
> You get linker errors if the atomics aren't actually implemented for that arch.
> I ran into this when I wrote the Nios II port.

Indeed! Atomic ops for ARC were only added in the end of 2015 (https://github.com/foss-for-synopsys-dwc-arc-processors/gcc/commit/f5599934783a78e3e89c69a88c6cc2fe741c7afa) and made it to arc-2016.03 release which was published on December 9-th 2016 (https://github.com/foss-for-synopsys-dwc-arc-processors/toolchain/releases/tag/arc-2016.09-release). which was much later than the first commit in Zephyr git repo :)

Mystery solved!